core/src/Kokkos_Array.hpp

```diff
@@ -320,6 +320,23 @@ struct Array<T, KOKKOS_INVALID_INDEX, Array<>::strided> {
       : m_elem(arg_ptr), m_size(arg_size), m_stride(arg_stride) {}
 };
 
+template <typename T, size_t N>
+KOKKOS_INLINE_FUNCTION constexpr bool operator==(const Array<T, N, void>& lhs,
+                                                 const Array<T, N, void>& rhs) {
+  // The cast to int is necessary to avoid a warning about a pointless
+  // comparison with zero in the case that an unsigned type is used.
+  for (int i = 0; i < (int)(lhs.size()); ++i) {
```

Comment:
Do you still get the pointless comparison warning if you use `!=` instead of `<`?  `!=` is more idiomatic C++ anyway.

If you still have to cast, use `static_cast` instead of the C-style cast.

Also, it really should be a hidden `friend` instead of a free function in the namespace.

Reply:
Hi @nliber. Thanks for the review and suggestions! And sorry for replying with some delay.

- The pointless comparison warning disappears if I use `!=` instead of `<`. That's interesting. For now, I've kept the use of `<` and the cast to remain consistent with what was done in PR #6423.
- I've replaced the C-style cast with a `static_cast`.
- I'm not sure how to best make it work as a hidden `friend`. An issue is that the `Kokkos::Array` has several specialisations. And we want the equality operator to work with both the specialisation for a fixed (non-zero) size and the one for the fixed zero size. So, if I move the equality operator inside the class definition, I think I'll have to define it twice. Also, because the equality operator uses public member functions, I think it wouldn't need to be a friend.

Reply:
Hidden `friend`s are about reducing the size of overload sets (faster compiles) by only using ADL (argument dependent lookup) and avoiding accidental implicit conversions.  It is conflated into the straightforward usage of `friend` (granting access to non-public members of a class).

Anthony Williams has a good article on hidden friends at [https://www.justsoftwaresolutions.co.uk/cplusplus/hidden-friends.html](https://www.justsoftwaresolutions.co.uk/cplusplus/hidden-friends.html).

Here are some examples:

```c++
class X {
    // hidden friend
    friend bool operator==(X const&, X const&) { return true; }
};
// Not hidden friend
inline bool operator!=(X const& lhs, X const& rhs) { return !(lhs == rhs); }

struct Y {
    operator X() const { return X(); }  // implicit conversion to X
};

X x;
Y y;
assert(  x == x );
assert(!(x != x));
assert(  x == y );
assert(!(x != y));
assert(  y == x );
assert(!(y != x));
assert(  y == y );  // error: no match for 'operator==' (operand types are 'Y' and 'Y')
assert(!(y != y));
```

```c++
class X {
    // hidden friend
    friend bool operator==(X const&, X const&) { return true; }
};
// Not hidden friend
inline bool operator!=(X const& lhs, X const& rhs) { return !(lhs == rhs); }

struct Y {
    operator X() const { return X(); }  // implicit conversion to X
};

X x;
Y y;
  operator==(x, x);
::operator==(x, x); // error: no member named 'operator==' in the global namespace
  operator!=(x, x);
::operator!=(x, x);
```

To be a hidden `friend`, it has to be the first declaration of a function and that declaration must contain the definition.  You'll have to repeat it in every specialization.

It also needs to be a `friend` because that is the only way to declare a non-member function inside a class.